doom/.config/doom/config.org
#+title: Config
#+AUTHOR: Praneeth Priyankara

* Doom Emacs Cofiguration

*** General Settings

- Setting the theme
#+begin_src emacs-lisp :tangle config.el
(use-package doom-themes
  :ensure t
  :config
  (load-theme 'doom-pine t)
  (doom-themes-visual-bell-config))
(setq doom-theme 'doom-pine)
#+end_src

- Unsetting the rainbow mode
  #+begin_src emacs-lisp :tangle config.el
(rainbow-mode -1)
  #+end_src

- Setting the font
#+begin_src emacs-lisp :tangle config.el
(setq doom-font (font-spec :family "IBM Plex Mono"  :size 16 :weight 'semi-bold)
      doom-variable-pitch-font (font-spec :family "Liga SFMono Nerd Font" :size 13 :weight 'regular))
#+end_src

- Show line numbers
#+begin_src emacs-lisp :tangle config.el
(setq display-line-numbers-type t)
#+end_src

- Enabling the clipboard
#+begin_src emacs-lisp :tangle config.el
(setq select-enable-clipboard t)
#+end_src

- Parenthesis highlighting
#+begin_src emacs-lisp :tangle config.el
(show-paren-mode 1)
#+end_src

- Global buffer reload
  #+begin_src emacs-lisp :tangle config.el
(global-auto-revert-mode 1)
  #+end_src

- Preventing killing the window
  - Preventing =:q= killing the window and exit only the buffer
#+begin_src emacs-lisp :tangle config.el
(after! evil
  (evil-ex-define-cmd "q" #'kill-current-buffer))
#+end_src

  - Preventing =:wq= killing the active window
#+begin_src emacs-lisp :tangle config.el
(after! evil
  (evil-ex-define-cmd "wq" #'kill-buffer--possibly-save))
#+end_src

  - Preventing the =Z Z= exiting the emacs
#+begin_src emacs-lisp :tangle config.el
(after! evil
  (define-key evil-normal-state-map (kbd "ZZ") #'save-buffer))
#+end_src

- Right click to copy to clipboard
#+begin_src emacs-lisp :tangle config.el
(map! "<mouse-3>" #'clipboard-kill-ring-save)
#+end_src

- Setting the the editor config
  #+begin_src emacs-lisp :tangle config.el
(editorconfig-mode 1)
  #+end_src

- Setting GPG entries to use emacs
  #+begin_src emacs-lisp :tangle config.el
(setq epg-pinentry-mode 'loopback)
  #+end_src

- Adding the scripts directory to path
#+begin_src emacs-lisp :tangle config.el
(add-to-list 'load-path "~/.config/doom/Scripts/")
#+end_src

- Sourcing the run.el
#+begin_src emacs-lisp :tangle config.el
(after! eshell
  (load (concat doom-user-dir "Scripts/run.el")))
#+end_src

- Word wrapping at startup
  #+begin_src emacs-lisp :tangle config.el
(add-hook! 'after-change-major-mode-hook #'+word-wrap-mode)
  #+end_src

*** Setting the Sinhala Language

- Setting the Sinhala language with fonts
#+begin_src emacs-lisp :tangle config.el
(set-fontset-font t 'sinhala (font-spec :family "Noto Sans Sinhala"))
#+end_src

- Setting block sizes for Sinhala language (This is needed to show the font shapes)
#+begin_src emacs-lisp :tangle config.el
(set-char-table-range composition-function-table
                      '(#x0D80 . #x0DFF)
                      (list (vector "[\x0D80-\x0DFF]+" 0 'font-shape-gstring)))
#+end_src

*** Org Mode

- Setting the Org directory
#+begin_src emacs-lisp :tangle config.el
(setq org-directory "~/org/")
#+end_src

- Code block indentation setting
#+begin_src emacs-lisp :tangle config.el
(after! dirvish
  (setq dirvish-attributes
        '(vc-state-icon file-slots icons file-size)))
#+end_src

*** Dired

Dired is the emacs' file manager. It is accessible with either =M-x dired= or =C-x d= in doom emacs.

- Setting to copy the path of the other pane to minibuffer, when in the dual pane mode. Dual pane mode is accessible with =C-x 3=.
#+begin_src emacs-lisp :tangle config.el
(setq dired-dwim-target t)
#+end_src

*** VTerm

VTerm is a terminal emulator for emacs.

- Start the shell with Fish shell
#+begin_src emacs-lisp :tangle config.el
(setq vterm-shell "/usr/bin/fish")
#+end_src

- Display Line Numbers
#+begin_src emacs-lisp :tangle config.el
(add-hook 'vterm-mode-hook (lambda () (display-line-numbers-mode -1)))
#+end_src

** Keybindings
**** Basic Bindings
- opening the vterm in the current directory.
#+begin_src emacs-lisp :tangle config.el
(map! :leader
      :desc "Opening VTerm" "v t" #'+vterm/here)
#+end_src

- opening the eshell
#+begin_src emacs-lisp :tangle config.el
(map! :leader
      :desc "Opening Eshell" "v e" #'eshell)
#+end_src


** Language Specific configuartions
**** JavaScript
#+begin_src emacs-lisp :tangle config.el
(after! org
  (org-babel-do-load-languages
   'org-babel-do-load-languages
   '((js . t))))
#+end_src

#+begin_src emacs-lisp :tangle config.el
(after! eglot
  (add-to-list 'eglot-server-programs
               '((typescript-tsx-mode typescript-mode js-mode js-ts-mode tsx-ts-mode)
                 . ("/home/priyankara/.config/nvm/versions/node/v26.2.0/bin/tsc" "--lsp" "--stdio"))))
#+end_src

**** MJML
#+begin_src emacs-lisp :tangle config.el
(add-to-list 'auto-mode-alist '("\\.mjml\\'" . web-mode))
(setq web-mode-content-types-alist '(("xml" . "\\.mjml\\'")))
#+end_src

**** YAML
#+begin_src emacs-lisp :tangle config.el
(org-babel-do-load-languages
 'org-babel-load-languages
 '((yaml . t)
   (shell . t)))
#+end_src

** Tool Specific Configurations

**** Jira
#+begin_src emacs-lisp :tangle config.el
(use-package! jira
  :config
  (setq jira-base-url "https://simplyxity.atlassian.net"
        jira-api-version 3))
#+end_src


doom/.config/doom/config.el
(use-package doom-themes
  :ensure t
  :config
  (load-theme 'doom-pine t)
  (doom-themes-visual-bell-config))
(setq doom-theme 'doom-pine)

(rainbow-mode -1)

(setq doom-font (font-spec :family "IBM Plex Mono"  :size 16 :weight 'semi-bold)
      doom-variable-pitch-font (font-spec :family "Liga SFMono Nerd Font" :size 13 :weight 'regular))

(setq display-line-numbers-type t)

(setq select-enable-clipboard t)

(show-paren-mode 1)

(global-auto-revert-mode 1)

(after! evil
  (evil-ex-define-cmd "q" #'kill-current-buffer))

(after! evil
  (evil-ex-define-cmd "wq" #'kill-buffer--possibly-save))

(after! evil
  (define-key evil-normal-state-map (kbd "ZZ") #'save-buffer))

(map! "<mouse-3>" #'clipboard-kill-ring-save)

(editorconfig-mode 1)

(setq epg-pinentry-mode 'loopback)

(add-to-list 'load-path "~/.config/doom/Scripts/")

(after! eshell
  (load (concat doom-user-dir "Scripts/run.el")))

(add-hook! 'after-change-major-mode-hook #'+word-wrap-mode)

(set-fontset-font t 'sinhala (font-spec :family "Noto Sans Sinhala"))

(set-char-table-range composition-function-table
                      '(#x0D80 . #x0DFF)
                      (list (vector "[\x0D80-\x0DFF]+" 0 'font-shape-gstring)))

(setq org-directory "~/org/")

(after! dirvish
  (setq dirvish-attributes
        '(vc-state-icon file-slots icons file-size)))

(setq dired-dwim-target t)

(setq vterm-shell "/usr/bin/fish")

(add-hook 'vterm-mode-hook (lambda () (display-line-numbers-mode -1)))

(map! :leader
      :desc "Opening VTerm" "v t" #'+vterm/here)

(map! :leader
      :desc "Opening Eshell" "v e" #'eshell)

(after! org
  (org-babel-do-load-languages
   'org-babel-do-load-languages
   '((js . t))))

(after! eglot
  (add-to-list 'eglot-server-programs
               '((typescript-tsx-mode typescript-mode js-mode js-ts-mode tsx-ts-mode)
                 . ("/home/priyankara/.config/nvm/versions/node/v26.2.0/bin/tsc" "--lsp" "--stdio"))))

(add-to-list 'auto-mode-alist '("\\.mjml\\'" . web-mode))
(setq web-mode-content-types-alist '(("xml" . "\\.mjml\\'")))

(org-babel-do-load-languages
 'org-babel-load-languages
 '((yaml . t)
   (shell . t)))

(use-package! jira
  :config
  (setq jira-base-url "https://simplyxity.atlassian.net"
        jira-api-version 3))
